packages/@tinacms/datalayer/src/database/bridge/isomorphic.ts

```diff
@@ -0,0 +1,541 @@
+          oid: existingOid,
+          cache: this.cache,
+        })
+
+        const updatedTree = treeResult.tree.filter(
+          (value) => value.path !== nodePath
+        )
+
+        // The folder is empty after removing the file, so we remove the entire folder
+        if (updatedTree.length === 0) {
+          ptr -= 1
+          continue
+        }
+
+        const updatedTreeOid = await git.writeTree({
+          ...this.isomorphicConfig,
+          tree: updatedTree,
+        })
+
+        const updatedRootTreeOid = await this.updateTreeHierarchy(
+          existingOid,
+          updatedTreeOid,
+          pathParts[ptr - 1],
+          'tree',
+          pathEntries.slice(0, ptr - 1),
+          pathParts.slice(0, ptr - 1)
+        )
+
+        await this.commitTree(updatedRootTreeOid, ref)
+
+        break
+      } else {
+        throw new GraphQLError(
+          `Unable to resolve path: ${filepath}`,
+          null,
+          null,
+          null,
+          null,
+          null,
+          { status: 404 }
+        )
+      }
+    }
+
+    await this.onDelete(filepath)
+  }
+
+  private qualifyPath(filepath: string): string {
+    return this.relativePath ? `${this.relativePath}/${filepath}` : filepath
+  }
+
+  private unqualifyPath(filepath: string): string {
+    return this.relativePath
+      ? filepath.slice(this.relativePath.length + 1)
+      : filepath
+  }
+
+  public async get(filepath: string) {
+    const ref = await this.getRef()
+    const oid = await git.resolveRef({
+      ...this.isomorphicConfig,
+      ref,
+    })
+
+    const { blob } = await git.readBlob({
+      ...this.isomorphicConfig,
+      oid,
+      filepath: this.qualifyPath(filepath),
+      cache: this.cache,
+    })
+
+    return Buffer.from(blob).toString('utf8')
+  }
+
+  public async putConfig(filepath: string, data: string) {
+    await this.put(filepath, data)
+  }
+
+  public async put(filepath: string, data: string) {
+    const ref = await this.getRef()
+    const { pathParts, pathEntries } = await this.resolvePathEntries(
+      this.qualifyPath(filepath),
+      ref
+    )
+
+    const blobUpdate = toUint8Array(Buffer.from(data))
+
+    let existingOid
+    const leafEntry = pathEntries[pathEntries.length - 1]
+    const nodePath = pathParts[pathParts.length - 1]
+    if (leafEntry) {
+      existingOid = await leafEntry.oid()
+      const hash = await git.hashBlob({ object: blobUpdate })
+      if (hash.oid === existingOid) {
+        await this.onPut(filepath, data)
+        return // no changes - exit early
+      }
+    }
+
+    const updatedOid = await git.writeBlob({
+      ...this.isomorphicConfig,
+      blob: blobUpdate,
+    })
+
+    const updatedRootSha = await this.updateTreeHierarchy(
+      existingOid,
+      updatedOid,
+      nodePath,
+      'blob',
+      pathEntries.slice(0, pathEntries.length - 1),
+      pathParts.slice(0, pathParts.length - 1)
+    )
+
+    await this.commitTree(updatedRootSha, ref)
```

Comment:
Can you explain at a high level what's going in the `put` function? I think I understand it to some extend but was confused by the output I saw (commented with the screenshots). 

In general I guess we'll need to nail down what it means to `put` in this context. Is that a "change not staged for commit", a "staged change", or a commit? How do you see that working?